### PR TITLE
Update Rails to version 7.2.2

### PR DESCRIPTION
**Pull request by mjankowski**

### CODE REVIEW

**Comment by renchap** [comment]
We will need to remember to remove our AR patches that will be introduced by https://github.com/mastodon/mastodon/pull/29889

**Comment by mjankowski** [comment]
Yup, misc other notes:

- There's a deprecation warning generated from a setting (I just commented out for now) -- we should probably work through this PR or something similar - https://github.com/mastodon/mastodon/pull/30332 - first, before the 7.2 upgrade (because 7.2 will presumably have its own similar generated file).
- There's an error from database_cleaner, they already have PR for https://github.com/DatabaseCleaner/database_cleaner-active_record/pull/101
- I had to update our reblog/transacation/safe monkey patch, want to review this more closely
- There's some changed interaction between activejob/sidekiq/emails and the various spec helpers used around all those. I want to dig into this more and understand why the changes I made here were needed.
- I used this workaround - https://github.com/rails/rails/issues/50604#issuecomment-1880990392 - for an activerecord encryption config issue ... want to find something better here as well
- Fabricator changes - https://gitlab.com/fabrication-gem/fabrication/-/issues/47#note_1911713203
- Thought I found an email_spec issue - https://github.com/email-spec/email-spec/issues/227 - but now cant replicate

**Comment by mjankowski** [comment]
There is a Rails 7.2.0.beta1 pushed to rubygems now - https://rubygems.org/gems/rails/versions/7.2.0.beta1

Updated this PR to use the released beta gem instead of the github branch.

I have not yet run the `app:update` to get new config in place. Might do that locally and see if anything can be extracted ahead of time.

**Comment by mjankowski** [comment]
Updated to use 7.2.0beta2

Have pulled a few things out, will continue to do so.

**Comment by mjankowski** [comment]
The DatabaseCleaner Rails 7.2 PR was merged, now pointing to main there, and looks all passing on CI.

I may break those spec change PRs into even smaller isolated PRs, and run the app:update in here at this point ... though we definitely need the linked PR about 7.1 defaults dealt with before this is ready for review.

**Comment by mjankowski** [comment]
Rebased this again after those spec-change PRs were merged, and also ran `app:update`. Going to pull some of that out early as well.

It looks like both the generated config file changes and the new framework defaults changes here are far far smaller than the 7.0 or 7.1 upgrades were (although we're still on beta2, so we'll see).

**Comment by mjankowski** [comment]
Updated to 7.2.0.beta3 - things still passing, still see the test-prof deprecation output, but it has an open issue.

**Comment by mjankowski** [comment]
Updated rails to `7.2.0.rc1`

**Comment by renchap** [comment]
Rails 7.2.0 is now out, but we do not plan to update to it before the 4.3 release.

**Comment by mjankowski** [comment]
Updated to 7.2.0 release, updated title/desc, converted to ready for review -- but noted (and agreed) on waiting on post-4.3 release.

**Comment by mjankowski** [comment]
Some cleanup to happen here:

- We're pointing to an RC release of test-prof, awaiting a final tag/release there
- The nsa gem has a hard version lock to `< 7.2` for rails deps ... but we have a separate PR removing nsa gem. For now, I've pulled in the removal into this PR, will rebase once that other PR merges
- database_cleaner gem also pointing to github, waiting on release
- Should do a once over on our safe reblog insert monkey patch and make sure the earlier versions match the release version
- Should regenerate the app:update changes and confirm that the prerelease changes in this PR now match the release changes
- Could review/merge this https://github.com/mastodon/mastodon/pull/30579 first, to reduce diff size here

**Comment by mjankowski** [comment]
Rebased, updated to 7.2.1.

**Comment by mjankowski** [comment]
With 4.3 out - would be great to get a review on this one.

Since this PR, we now have 8.0 betas coming out.

Separately - getting the linked formatting-only PR merged would be useful to keep this one changes-only focused.

**Comment by mjankowski** [comment]
Rebased, and updated to remove the two patches that are included in 7.2

**Comment by mjankowski** [comment]
Pulled out another precursor thing similar to some past ones.

Updated to use just-released 7.2.1.1

**Comment by mjankowski** [comment]
Starting to build some needs-7.2 features branches...

WIP on `with_routing` - https://github.com/mjankowski/mastodon/compare/rails-7-2...mjankowski:mastodon:rails-7-2-with-routing

WIP on `touch_all` - https://github.com/mjankowski/mastodon/compare/rails-7-2...mjankowski:mastodon:rails-7-2-touch-all

File: app/models/concerns/status/safe_reblog_insert.rb
Comment (renchap) [inline app/models/concerns/status/safe_reblog_insert.rb:40]: I am not familiar with this code, but why are we taking `connection` as a method parameter, then calling `with_connection`, which seems to yield you a connection, and not use it?

File: config/environments/development.rb
Comment (renchap) [inline config/environments/development.rb:40]: This is now built-in?

File: config/environments/development.rb
Comment (renchap) [inline config/environments/development.rb:89]: Maybe we could enable this, it seems useful?

File: config/environments/test.rb
Comment (renchap) [inline config/environments/test.rb:23]: We are using our own middleware for serving static files, should we keep this here as its not used (from what I understand)?

File: app/models/concerns/status/safe_reblog_insert.rb
Comment (mjankowski) [reply app/models/concerns/status/safe_reblog_insert.rb:40]: Why are we doing this? -- I believe this is just updated to reflect what changed in the actual rails class (including this non-use, see comment link).

Why is it like that? -- no idea.

File: config/environments/development.rb
Comment (mjankowski) [reply config/environments/development.rb]: Agreed! For first pass I was regening and not thinking too hard on new options. Will enable.

File: config/environments/development.rb
Comment (mjankowski) [reply config/environments/development.rb:40]: ...I forget the exact changes in each of 7.2 vs 8.0, but yes I think something like this has been made a default config. Production still has the more explicit assignment so that it includes the configured log_formatter.

**Comment by mjankowski** [comment]
Rebased this after the linked formatting-only PR merged, and then re-ran the `app:update` process, which yielded a few more pre-merge todos here:

- There are a few more grammar/wording only changes in the final release, will pull those out
- I want to re-examine the reblog monkey patch and make sure it's still valid against the final release (I had it against a beta initially). Probably fine but will double check.
- The generators added a rubocop optimization which I included here after rebase ... which had the unintended side effect of fixing what were actually some hidden violations (discussed in various other PRs, there's a path/loading/config issue). Some of these already have open PRs, will handle the rest separately.

I don't think any of those impact the ability to do the rails upgrade (which I think is fine?), but just stuff that came up during the process here.

File: Gemfile
Comment (mjankowski) [inline Gemfile:65]: The `httpclient` gem needs this but does not depend on it.

Separately, that gem looks very unmaintained. I see there's a shopify fork. Maybe something to check out.

File: app/models/concerns/status/safe_reblog_insert.rb
Comment (mjankowski) [inline app/models/concerns/status/safe_reblog_insert.rb:19]: I think it makes sense to keep this annotated as we update it.

Another approach I've seen is to put some sort of initializer version check in, and raise an error on any updates, which sort of forces the issue of manual verifying compatibility on every update.

File: config/application.rb
Comment (mjankowski) [inline config/application.rb:56]: These are both merged and in rails now.

File: config/application.rb
Comment (mjankowski) [inline config/application.rb:62]: Going directly to this instead of the new_framework_defaults file, because its very minimal and has nothing for us to step through really (or rather, everything is either desirable or not relevant) - https://github.com/rails/rails/blob/v7.2.1.1/railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_2.rb.tt

File: config/environments/development.rb
Comment (mjankowski) [inline config/environments/development.rb:42]: This style config remains in place for production, but the default is basically this already for dev mode.

File: config/environments/development.rb
Comment (mjankowski) [inline config/environments/development.rb:91]: Seems nice!

File: config/environments/production.rb
Comment (mjankowski) [inline config/environments/production.rb:20]: This is just moved below to top level config, not removed.

I think this configuration was setting the asset host for controller only but not also for mailers ... and I'm assuming we want it for both (hence move to top level). If that's NOT true, I'll revert.

File: config/environments/production.rb
Comment (mjankowski) [inline config/environments/production.rb:52]: This moves below, the to_sym is handled by lib now.

File: config/environments/test.rb
Comment (mjankowski) [inline config/environments/test.rb:22]: I believe this is no longer relevant/needed based on what the propshaft/webpack interaction is.

File: config/environments/test.rb
Comment (mjankowski) [inline config/environments/test.rb:46]: I think this is safe given test env, but could use more review ... we have the `1_hosts` intializer doing some variation on this.

Sort of related - various ongoing discussion about aligning the diff portions of the spec suite to each other re: hostnames/ports/etc.

File: config/environments/test.rb
Comment (mjankowski) [inline config/environments/test.rb:61]: Winds up set this way w/out this here.

**Comment by mjankowski** [comment]
Added various inline annotation on the state of things here ... prob ready for review.

File: app/models/concerns/status/safe_reblog_insert.rb
Comment (renchap) [reply app/models/concerns/status/safe_reblog_insert.rb:40]: Good point, I missed that it was like that in the upstream file

File: Gemfile
Comment (renchap) [reply Gemfile:65]: I am not sure where we use this gem, maybe this should indeed be replaced entirely?

Comment (mjankowski) [reply Gemfile:65]: It's an indirect dependency, pulled in by a few gems.

File: config/application.rb
Comment (mjankowski) [inline config/application.rb:67]: Will be set to this by default.

**Comment by mjankowski** [comment]
Rebased, bumped to 7.2.1.2

**Review by oneiros — DISMISSED** [review]

**Comment by mjankowski** [comment]
Rebased, bumped to 7.2.2

**Review by oneiros — DISMISSED** [review]

File: config/environments/production.rb
Comment (mjankowski) [inline config/environments/production.rb:32]: This is not new, just pulling down existing config to new location.

**Comment by mjankowski** [comment]
I see this is approved but not merged -- anything else needed from me here, or just waiting until someone is excited about merging?

**Comment by oneiros** [comment]
> I see this is approved but not merged -- anything else needed from me here, or just waiting until someone is excited about merging?

More the latter I guess. Sorry for the confusion.

**Review by oneiros — APPROVED** [review]

**Comment by shleeable** [comment]
"12 conversations must be resolved before merging."

That's new to me

**Comment by mjankowski** [comment]
> "12 conversations must be resolved before merging."

I believe when you "start a review" this happens. If you just leave comments there's no dependency on resolving first.

I sometimes (by accident!) start reviews instead of just commenting. When that happens I like to take myself aside, take a long hard look in the mirror, forgive myself, and get back in the game.

**Pull request merged**
